File: shared-protocol/python/src/jarvis_protocol/schemas.py
"""Loads the canonical JSON Schemas and builds Draft 2020-12 validators.

The schemas in ``shared-protocol/schema`` are the single source of truth. This
module locates that directory, registers every schema by ``$id`` so that
cross-file ``$ref``s resolve, and exposes cached validators.

Resolution order for the schema directory:
  1. ``$JARVIS_PROTOCOL_SCHEMA_DIR`` env var, if it points at a dir with the schemas.
  2. The nearest ``schema/`` directory walking up from this file (works for an
     editable/source checkout).
"""

from __future__ import annotations

import json
import os
from functools import lru_cache
from pathlib import Path
from typing import Dict, Optional

from jsonschema import Draft202012Validator
from referencing import Registry, Resource
from referencing.jsonschema import DRAFT202012

from .catalog import TYPE_TO_SCHEMA


def _find_schema_dir() -> Path:
    candidates = []
    env = os.environ.get("JARVIS_PROTOCOL_SCHEMA_DIR")
    if env:
        candidates.append(Path(env))
    here = Path(__file__).resolve()
    candidates.extend(parent / "schema" for parent in here.parents)
    for cand in candidates:
        if (cand / "envelope.schema.json").is_file():
            return cand
    raise RuntimeError(
        "Could not locate shared-protocol/schema. "
        "Set the JARVIS_PROTOCOL_SCHEMA_DIR environment variable to its path."
    )


#: Absolute path to the canonical schema directory.
SCHEMA_DIR: Path = _find_schema_dir()


@lru_cache(maxsize=1)
def load_schemas() -> Dict[str, dict]:
    """Return ``{filename: schema}`` for every ``*.schema.json`` in SCHEMA_DIR."""
    schemas: Dict[str, dict] = {}
    for path in sorted(SCHEMA_DIR.glob("*.schema.json")):
        schemas[path.name] = json.loads(path.read_text(encoding="utf-8"))
    if "envelope.schema.json" not in schemas:  # pragma: no cover - sanity
        raise RuntimeError(f"envelope.schema.json missing from {SCHEMA_DIR}")
    return schemas


@lru_cache(maxsize=1)
def _registry() -> Registry:
    resources = [
        (schema["$id"], Resource.from_contents(schema, default_specification=DRAFT202012))
        for schema in load_schemas().values()
    ]
    return Registry().with_resources(resources)


@lru_cache(maxsize=1)
def envelope_validator() -> Draft202012Validator:
    return Draft202012Validator(load_schemas()["envelope.schema.json"], registry=_registry())


@lru_cache(maxsize=None)
def payload_validator(type_name: str) -> Optional[Draft202012Validator]:
    """Validator for a message ``type``'s payload, or ``None`` if unknown."""
    filename = TYPE_TO_SCHEMA.get(type_name)
    if filename is None:
        return None
    return Draft202012Validator(load_schemas()[filename], registry=_registry())


__all__ = [
    "SCHEMA_DIR",
    "load_schemas",
    "envelope_validator",
    "payload_validator",
]

File: shared-protocol/typescript/src/schemas.ts
/**
 * Loads the canonical JSON Schemas and builds Ajv (draft 2020-12) validators.
 *
 * The schemas in `shared-protocol/schema` are the single source of truth. We
 * locate that directory (env override or upward search), register every schema
 * by `$id` so cross-file `$ref`s resolve, and expose cached validators.
 */
import { existsSync, readFileSync, readdirSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { Ajv2020, type ErrorObject, type ValidateFunction } from "ajv/dist/2020.js";
import { TYPE_TO_SCHEMA } from "./catalog.js";

/**
 * Locate the canonical schema directory. Exported (but not re-exported from the
 * package index) so tests can exercise the env / upward-search / not-found paths.
 * `opts.env` / `opts.startDir` override the defaults for testing.
 */
export function findSchemaDir(opts: { env?: string; startDir?: string } = {}): string {
  const candidates: string[] = [];
  const env = opts.env ?? process.env.JARVIS_PROTOCOL_SCHEMA_DIR;
  if (env) candidates.push(env);
  let dir = opts.startDir ?? dirname(fileURLToPath(import.meta.url));
  for (;;) {
    candidates.push(join(dir, "schema"));
    const parent = dirname(dir);
    if (parent === dir) break;
    dir = parent;
  }
  for (const cand of candidates) {
    if (existsSync(join(cand, "envelope.schema.json"))) return cand;
  }
  throw new Error(
    "Could not locate shared-protocol/schema. Set JARVIS_PROTOCOL_SCHEMA_DIR to its path.",
  );
}

/** Absolute path to the canonical schema directory. */
export const SCHEMA_DIR = findSchemaDir();

let _schemas: Record<string, Record<string, unknown>> | null = null;

/** Return `{ filename: schema }` for every `*.schema.json` in SCHEMA_DIR. */
export function loadSchemas(): Record<string, Record<string, unknown>> {
  if (_schemas) return _schemas;
  const out: Record<string, Record<string, unknown>> = {};
  for (const file of readdirSync(SCHEMA_DIR)) {
    if (file.endsWith(".schema.json")) {
      out[file] = JSON.parse(readFileSync(join(SCHEMA_DIR, file), "utf-8"));
    }
  }
  _schemas = out;
  return out;
}

let _ajv: Ajv2020 | null = null;

export function ajv(): Ajv2020 {
  if (_ajv) return _ajv;
  const instance = new Ajv2020({ allErrors: true, strict: false });
  for (const schema of Object.values(loadSchemas())) {
    instance.addSchema(schema);
  }
  _ajv = instance;
  return instance;
}

export function envelopeValidator(): ValidateFunction {
  const fn = ajv().getSchema("https://jarvisvr.dev/schema/envelope.schema.json");
  /* c8 ignore next -- defensive: the envelope schema is always registered */
  if (!fn) throw new Error("envelope schema not registered");
  return fn;
}

export function payloadValidator(type: string): ValidateFunction | null {
  const file = TYPE_TO_SCHEMA[type];
  if (!file) return null;
  const schema = loadSchemas()[file];
  const id = schema["$id"] as string;
  /* c8 ignore next -- getSchema always resolves a registered $id for a known type */
  return ajv().getSchema(id) ?? null;
}

export type { ErrorObject, ValidateFunction };
